hal/src/sercom/v2/uart/mod.rs

```diff
@@ -0,0 +1,1005 @@
+//!   must be enabled.
+//!
+//! [`enable`]: Config::enable
+//! [`disable`]: Uart::disable
+//! [`reconfigure`]: Uart::reconfigure
+//! [`bsp_pins`]: crate::bsp_pins
+//! [`Pin`]: crate::gpio::v2::pin::Pin
+//! [`PinId`]: crate::gpio::v2::pin::PinId
+//! [`PinMode`]: crate::gpio::v2::pin::PinMode
+//! [`split`]: Uart::split
+//! [`join`]: Uart::join
+//! [`NoneT`]: crate::typelevel::NoneT
+//! [`serial::Write`]: embedded_hal::serial::Write
+//! [`serial::Read`]: embedded_hal::serial::Read
+#![cfg_attr(
+    feature = "dma",
+    doc = "
+# Using UART with DMA
+
+This HAL includes support for DMA-enabled UART transfers. [`Uart`]
+implements the DMAC [`Buffer`]
+trait. The provided [`send_with_dma`] and
+[`receive_with_dma`] build and begin a
+[`dmac::Transfer`], thus starting the UART
+in a non-blocking way. Optionally, interrupts can be enabled on the provided
+[`Channel`]. Note that the `dma` feature must
+be enabled. Please refer to the [`dmac`](crate::dmac) module-level
+documentation for more information.
+
+```
+/// Assume channel0 and channel1 are configured `dmac::Channel`s,
+/// rx is a Uart<C, RxDuplex>, and tx is a Uart<C, TxDuplex>.
+
+/// Create data to send
+let tx_buffer: [u8; 50] = [0xff; 50];
+let rx_buffer: [u8; 100] = [0xab; 100];
+
+/// Launch transmit transfer
+let tx_dma = tx.send_with_dma(&mut tx_buffer, channel0, ());
+
+/// Launch receive transfer
+let rx_dma = rx.receive_with_dma(&mut rx_buffer, channel1, ());
+
+/// Wait for transfers to complete and reclaim resources
+let (chan0, tx_buffer, tx) = tx_dma.wait();
+let (chan1, rx, rx_buffer) = rx_dma.wait();
+```
+
+[`Buffer`]: crate::dmac::transfer::Buffer
+[`send_with_dma`]: Uart::send_with_dma
+[`receive_with_dma`]: Uart::receive_with_dma
+[`dmac::Transfer`]: crate::dmac::Transfer
+[`Channel`]: crate::dmac::channel::Channel
+[`dmac`]: crate::dmac
+
+"
+)]
+
+#[cfg(any(feature = "samd11", feature = "samd21"))]
+#[path = "pads_thumbv6m.rs"]
+mod pads;
+
+#[cfg(feature = "min-samd51g")]
+#[path = "pads_thumbv7em.rs"]
+mod pads;
+
+pub use pads::*;
+
+mod reg;
+use reg::Registers;
+
+mod charsize;
+pub use charsize::*;
+
+mod flags;
+pub use flags::*;
+
+mod config;
+pub use config::*;
+
+pub mod impl_ehal;
+
+use crate::{sercom::v2::*, typelevel::Sealed};
+use core::{convert::TryInto, marker::PhantomData};
+use num_traits::AsPrimitive;
+
+//=============================================================================
+// Stop bits, parity, baud rate, bit order
+//=============================================================================
+
+/// Number of stop bits in a UART frame
+#[repr(u8)]
+#[derive(Debug, Clone, Copy)]
+pub enum StopBits {
+    /// 1 stop bit
+    OneBit = 0,
+    /// 2 stop bits
+    TwoBits = 1,
+}
+
+impl From<StopBits> for bool {
+    #[inline]
+    fn from(item: StopBits) -> bool {
+        match item {
+            StopBits::OneBit => false,
+            StopBits::TwoBits => true,
+        }
+    }
+}
```

Comment:
Do you think users will actually use these trait definitions at all? It seems like they are only relevant for setting the registers. In that case, they would only be used once within the HAL and nowhere else. Would it make sense to move them there?

Do you see any use cases I'm missing?

Reply:
No, you're right. Removed them.